### DISCUSSION TITLE
Why does kitty run bash (my user shell) with the --posix parameter?

### DISCUSSION THREAD

**Discussion by Filip62** [kovidgoyal/kitty#4807] [Q&A] [answered]
Hello.

I was scrolling through htop and noticed that kitty starts `bash --posix` instead of just pure `bash` for some reason.

In my config I haven't changed the shell option, it looks like the default `shell .`

In /etc/passwd I have `filip:x:1000:1000::/home/filip:/bin/bash`

![image](https://user-images.githubusercontent.com/58399088/157434726-d0850203-7802-4b70-a904-b4e10bb41cb4.png)

Alacritty and xterm with default settings open normal `bash`.

I'm not sure what's going on. Any suggestions are appreciated, thank You.

Kitty 0.24.4, Arch linux, bspwm.

**Comment by page-down** [comment, marked as answer]
kitty provides shell integration without modifying rc files by running bash POSIX mode.
POSIX mode will be disabled after running, so there is nothing to worry about.

https://sw.kovidgoyal.net/kitty/shell-integration/#how-it-works

**Reply by Filip62** [reply]
I see, thanks.

**Comment by infogrind** [comment]
> POSIX mode will be disabled after running, so there is nothing to worry about.

The documentation at https://sw.kovidgoyal.net/kitty/shell-integration/#how-it-works says

> The loading of the startup files is done by the integration script, after disabling POSIX mode. From the perspective of those scripts there should be no difference to running vanilla bash.

That is not what I observe. In my `.profile`, I am sourcing some scripts to enable bash auto completion. Some of these use bash's `<( )` syntax. When starting kitty, I get the following error:

```
-bash: /Library/GoogleCorpSupport/srcfs/shell_completion/p4: line 25: syntax error near unexpected token `<'
-bash: /Library/GoogleCorpSupport/srcfs/shell_completion/p4: line 25: `    <( ... `
```
The same error doesn't occur if I open my normal Mac OS terminal.

I can also reproduce the issue simply by starting kitty and typing
```
echo <(ls)
```
Actual output:
```
-bash: syntax error near unexpected token `('
```
Expected output: something like 
```
/dev/fd/63
```

When I explicity start a new bash instance (without the `--posix` flag), the command succeeds.

I am running kitty 0.25.2 on Mac OS Monterey 12.4.

**Comment by kovidgoyal** [comment]
On Mon, Jun 20, 2022 at 07:40:30AM -0700, Marius Kleiner wrote:
> That is not what I observe. In my `.profile`, I am sourcing some scripts to enable bash auto completion. Some of these use bash's `<( )` syntax. When starting kitty, I get the following error:
> 
> ```
> -bash: /Library/GoogleCorpSupport/srcfs/shell_completion/p4: line 25: syntax error near unexpected token `<'
> -bash: /Library/GoogleCorpSupport/srcfs/shell_completion/p4: line 25: `    <( ... `
> ```
> The same error doesn't occur if I open my normal Mac OS terminal.
> 
> I can also reproduce the issue simply by starting kitty and typing
> ```
> echo <(ls)
> ```
> Actual output:
> ```
> -bash: syntax error near unexpected token `('
> ```
> Expected output: something like 
> ```
> /dev/fd/63
> ```
> 
> When I explicity start a new bash instance (without the `--posix` flag), the command succeeds.
> 
> I am running kitty 0.25.2 on Mac OS Monterey 12.4.

Does not reproduce for me.

kitty --config=NONE -o shell=bash

echo <(ls)
/dev/fd/63


I am guessing you have multiple bashes on your system and a different
one is being executed from the one you expect.

**Reply by infogrind** [reply]
Thanks Kovid for replying so quickly (I got to love Kitty within just a single weekend 😍).

> kitty --config=NONE -o shell=bash

Do I have to configure in some way how to start Kitty? All I did was download the DMG from the Github releases page and run it from the Applications folder, nothing more.

> I am guessing you have multiple bashes on your system and a different
one is being executed from the one you expect.

I tried to validate this hypothesis. First by getting the PID of the running bash instance:

```
~$ ps
  PID TTY           TIME CMD
 1558 ttys001    0:00.03 -bash --posix
```

And then applying [this method](https://unix.stackexchange.com/a/27896) to find the executable:
```
lsof -n | grep 1558 | grep ' txt'
```
This gives me
```
bash       1558 m*******  txt       REG                1,4    1326576 1152921500312779617 /bin/bash
```

I then did the same using the default Mac OS Terminal.app (after completely closing/quitting Kitty). There, I see for once that the `--posix` flag is not there:
```
$ ps
  PID TTY           TIME CMD
 2721 ttys000    0:00.02 -bash
```
And then
```
$ lsof -n | grep 2721 | grep ' txt'
bash       2721 m*******  txt       REG                1,4    1326576 1152921500312779617 /bin/bash
```

From this it looks like they both use the same binary.

**Comment by kovidgoyal** [comment]
On Mon, Jun 20, 2022 at 08:45:59AM -0700, Marius Kleiner wrote:
> Thanks Kovid for replying so quickly (I got to love Kitty within just a single weekend 😍).
> 
> > kitty --config=NONE -o shell=bash
> 
> Do I have to configure in some way how to start Kitty? All I did was download the DMG from the Github releases page and run it from the Applications folder, nothing more.

That is just a way of launching kitty with no config and shell set to
bash to aid reproduction.

> 
> > I am guessing you have multiple bashes on your system and a different
> one is being executed from the one you expect.
> 
> I tried to validate this hypothesis. First by getting the PID of the running bash instance:
> 
> ```
> ~$ ps
>   PID TTY           TIME CMD
>  1558 ttys001    0:00.03 -bash --posix
> ```

/bin/bash on macOS is limited to version 3.2 and does not support
a whole bunch of things. If you are really using that, then I suggest
you turn off shell_integration in kitty.conf and you will be fine.

**Reply by infogrind** [reply]
I suppose you have a point. And indeed, I did not plan to use Kitty with Bash. Using it with Fish now, and everything is OK.

Thanks again Kovid!

**Comment by alper** [comment]
Just for your reference, this is causing an issue say here: https://github.com/starship/starship/issues/5382

**Comment by kovidgoyal** [comment]
No, it isn't. Startup files are sourced with POSIX mode disabled, as
documented.
